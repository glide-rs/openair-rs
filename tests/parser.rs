use std::io::Cursor;

use indoc::indoc;
use insta::assert_debug_snapshot;
use openair::*;

#[test]
fn test_switzerland_fixture() {
    let airspace = include_bytes!("../example_data/Switzerland.txt");
    let mut cursor = Cursor::new(airspace);

    let spaces = parse(&mut cursor).unwrap();
    assert_debug_snapshot!(spaces.first().unwrap());

    let names = spaces.iter().map(|a| a.to_string()).collect::<Vec<_>>();
    assert_debug_snapshot!(names);
}

#[test]
fn test_germany_fixture() {
    let airspace = include_bytes!("../example_data/Germany.txt");
    let mut cursor = Cursor::new(airspace);

    let spaces = parse(&mut cursor).unwrap();
    assert_debug_snapshot!(spaces.first().unwrap());

    let names = spaces.iter().map(|a| a.to_string()).collect::<Vec<_>>();
    assert_debug_snapshot!(names);
}

#[test]
fn test_germany_border_fixture() {
    let airspace = include_bytes!("../example_data/Germany_Border.txt");
    let mut cursor = Cursor::new(airspace);

    let spaces = parse(&mut cursor).unwrap();
    assert_debug_snapshot!(spaces.first().unwrap());

    let names = spaces.iter().map(|a| a.to_string()).collect::<Vec<_>>();
    assert_debug_snapshot!(names);
}

/// Parse an airspace as generated by flyland.
#[test]
fn flyland_buochs() {
    let mut airspace = indoc! {"
        AC D
        AN BUOCHS Be CTR 119.625
        AL GND
        AH 12959 ft
        DP 46:57:13 N 008:27:52 E
        DP 46:57:46 N 008:30:41 E
        DP 46:57:55 N 008:28:40 E
        DP 46:58:28 N 008:27:56 E
        DP 46:57:13 N 008:27:52 E
        * n-Points: 5
    "}
    .as_bytes();
    let mut spaces = parse(&mut airspace).unwrap();
    assert_eq!(spaces.len(), 1);
    let space: Airspace = spaces.pop().unwrap();
    assert_eq!(space.name, "BUOCHS Be CTR 119.625");
    assert_eq!(space.lower_bound, Altitude::Gnd);
    assert_eq!(space.upper_bound, Altitude::FeetAmsl(12959));
    if let Geometry::Polygon { segments } = space.geom {
        assert_eq!(segments.len(), 5);
    } else {
        panic!("Unexpected enum variant");
    }
}

/// The order of bounds should not matter.
#[test]
fn inverted_bounds() {
    let mut a1 = indoc! {"
        AC D
        AN SOMESPACE
        AL GND
        AH 12959 ft
        DP 46:57:13 N 008:27:52 E
        DP 46:57:46 N 008:30:41 E
        *
    "}
    .as_bytes();
    let mut a2 = indoc! {"
        AC D
        AN SOMESPACE
        AH 12959 ft
        AL GND
        DP 46:57:13 N 008:27:52 E
        DP 46:57:46 N 008:30:41 E
        *
    "}
    .as_bytes();
    let space1 = parse(&mut a1).unwrap().pop().unwrap();
    let space2 = parse(&mut a2).unwrap().pop().unwrap();
    assert_eq!(space1, space2);
}

/// Variables can be defined multiple times.
#[test]
fn multi_variable() {
    let mut a = indoc! {"
        AC D
        AN SOMESPACE
        AL GND
        AH FL100
        V X=52:00:00N 013:00:00E
        V D=+
        DA 2,0,30
        V X=52:00:00N 013:00:00E
        V D=-
        DA 4,60,30
        *
    "}
    .as_bytes();
    let airspace = parse(&mut a).unwrap().pop().unwrap();
    assert_eq!(
        airspace.geom,
        Geometry::Polygon {
            segments: vec![
                PolygonSegment::ArcSegment(ArcSegment {
                    centerpoint: Coord {
                        lat: 52.0,
                        lng: 13.0
                    },
                    radius: 2.0,
                    angle_start: 0.0,
                    angle_end: 30.0,
                    direction: Direction::Cw,
                }),
                PolygonSegment::ArcSegment(ArcSegment {
                    centerpoint: Coord {
                        lat: 52.0,
                        lng: 13.0
                    },
                    radius: 4.0,
                    angle_start: 60.0,
                    angle_end: 30.0,
                    direction: Direction::Ccw,
                }),
            ],
        }
    );
}

/// Test AY/AF/AG records.
#[test]
fn extension_records() {
    let mut a = indoc! {"
        AC D
        AN SOMESPACE
        AL GND
        AH 100 ft AGL
        AY AWY
        AF 132.350
        AG Dutch Mil
        AX 1234
        A* custom extension
        V X=52:00:00 N 013:00:00 E
        DC 5
    "}
    .as_bytes();
    let airspace = parse(&mut a).unwrap().pop().unwrap();
    assert_eq!(airspace.type_, Some("AWY".to_string()));
    assert_eq!(airspace.frequency, Some("132.350".to_string()));
    assert_eq!(airspace.call_sign, Some("Dutch Mil".to_string()));
    assert_eq!(airspace.transponder_code, Some(1234));
}
