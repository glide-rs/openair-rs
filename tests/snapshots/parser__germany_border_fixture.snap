---
source: tests/parser.rs
expression: spaces.first().unwrap()
---
Airspace {
    name: "FIR Bremen 125.100",
    class: C,
    type_: None,
    lower_bound: FlightLevel(
        100,
    ),
    upper_bound: FlightLevel(
        245,
    ),
    geom: Polygon {
        segments: [
            Point(
                Coord {
                    lat: 55.0,
                    lng: 6.5,
                },
            ),
            Point(
                Coord {
                    lat: 55.0,
                    lng: 8.0,
                },
            ),
            Point(
                Coord {
                    lat: 55.06666666666667,
                    lng: 8.333333333333334,
                },
            ),
            Point(
                Coord {
                    lat: 55.06916666666667,
                    lng: 8.391944444444444,
                },
            ),
            Point(
                Coord {
                    lat: 55.06916666666667,
                    lng: 8.5,
                },
            ),
            Point(
                Coord {
                    lat: 54.91,
                    lng: 8.651388888888889,
                },
            ),
            Point(
                Coord {
                    lat: 54.91416666666667,
                    lng: 8.681666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 54.91222222222222,
                    lng: 8.698055555555555,
                },
            ),
            Point(
                Coord {
                    lat: 54.89694444444444,
                    lng: 8.729166666666666,
                },
            ),
            Point(
                Coord {
                    lat: 54.89666666666667,
                    lng: 8.755,
                },
            ),
            Point(
                Coord {
                    lat: 54.89972222222222,
                    lng: 8.7575,
                },
            ),
            Point(
                Coord {
                    lat: 54.913888888888884,
                    lng: 8.769444444444446,
                },
            ),
            Point(
                Coord {
                    lat: 54.915277777777774,
                    lng: 8.793611111111112,
                },
            ),
            Point(
                Coord {
                    lat: 54.913888888888884,
                    lng: 8.817777777777778,
                },
            ),
            Point(
                Coord {
                    lat: 54.91111111111111,
                    lng: 8.826666666666666,
                },
            ),
            Point(
                Coord {
                    lat: 54.902499999999996,
                    lng: 8.83388888888889,
                },
            ),
            Point(
                Coord {
                    lat: 54.89333333333333,
                    lng: 8.848611111111111,
                },
            ),
            Point(
                Coord {
                    lat: 54.89666666666667,
                    lng: 8.865277777777777,
                },
            ),
            Point(
                Coord {
                    lat: 54.90083333333333,
                    lng: 8.871666666666668,
                },
            ),
            Point(
                Coord {
                    lat: 54.90416666666667,
                    lng: 8.891388888888889,
                },
            ),
            Point(
                Coord {
                    lat: 54.90361111111111,
                    lng: 8.917222222222222,
                },
            ),
            Point(
                Coord {
                    lat: 54.901944444444446,
                    lng: 8.936666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 54.8975,
                    lng: 8.981944444444444,
                },
            ),
            Point(
                Coord {
                    lat: 54.88666666666666,
                    lng: 9.0025,
                },
            ),
            Point(
                Coord {
                    lat: 54.87916666666667,
                    lng: 9.029444444444445,
                },
            ),
            Point(
                Coord {
                    lat: 54.874722222222225,
                    lng: 9.089722222222223,
                },
            ),
            Point(
                Coord {
                    lat: 54.875,
                    lng: 9.120000000000001,
                },
            ),
            Point(
                Coord {
                    lat: 54.87361111111111,
                    lng: 9.142777777777777,
                },
            ),
            Point(
                Coord {
                    lat: 54.87277777777778,
                    lng: 9.154444444444445,
                },
            ),
            Point(
                Coord {
                    lat: 54.855000000000004,
                    lng: 9.21861111111111,
                },
            ),
            Point(
                Coord {
                    lat: 54.835,
                    lng: 9.256944444444445,
                },
            ),
            Point(
                Coord {
                    lat: 54.81916666666667,
                    lng: 9.275555555555556,
                },
            ),
            Point(
                Coord {
                    lat: 54.805277777777775,
                    lng: 9.274722222222223,
                },
            ),
            Point(
                Coord {
                    lat: 54.800555555555555,
                    lng: 9.289722222222222,
                },
            ),
            Point(
                Coord {
                    lat: 54.80222222222222,
                    lng: 9.2975,
                },
            ),
            Point(
                Coord {
                    lat: 54.80972222222222,
                    lng: 9.303611111111111,
                },
            ),
            Point(
                Coord {
                    lat: 54.80472222222222,
                    lng: 9.348611111111111,
                },
            ),
            Point(
                Coord {
                    lat: 54.81472222222222,
                    lng: 9.364444444444445,
                },
            ),
            Point(
                Coord {
                    lat: 54.836111111111116,
                    lng: 9.382222222222223,
                },
            ),
            Point(
                Coord {
                    lat: 54.84111111111111,
                    lng: 9.405277777777778,
                },
            ),
            Point(
                Coord {
                    lat: 54.847500000000004,
                    lng: 9.472222222222223,
                },
            ),
            Point(
                Coord {
                    lat: 54.765,
                    lng: 10.053611111111111,
                },
            ),
            Point(
                Coord {
                    lat: 54.74305555555556,
                    lng: 10.166666666666666,
                },
            ),
            Point(
                Coord {
                    lat: 54.7,
                    lng: 10.333333333333334,
                },
            ),
            Point(
                Coord {
                    lat: 54.65833333333333,
                    lng: 10.5,
                },
            ),
            Point(
                Coord {
                    lat: 54.65555555555555,
                    lng: 10.666666666666666,
                },
            ),
            Point(
                Coord {
                    lat: 54.65277777777778,
                    lng: 10.833333333333334,
                },
            ),
            Point(
                Coord {
                    lat: 54.644444444444446,
                    lng: 11.0,
                },
            ),
            Point(
                Coord {
                    lat: 54.60277777777778,
                    lng: 11.166666666666666,
                },
            ),
            Point(
                Coord {
                    lat: 54.55416666666667,
                    lng: 11.333333333333334,
                },
            ),
            Point(
                Coord {
                    lat: 54.5,
                    lng: 11.5,
                },
            ),
            Point(
                Coord {
                    lat: 54.46388888888889,
                    lng: 11.666666666666666,
                },
            ),
            Point(
                Coord {
                    lat: 54.44583333333333,
                    lng: 11.833333333333334,
                },
            ),
            Point(
                Coord {
                    lat: 54.45,
                    lng: 12.0,
                },
            ),
            Point(
                Coord {
                    lat: 54.45,
                    lng: 12.0,
                },
            ),
            Point(
                Coord {
                    lat: 54.2,
                    lng: 11.3,
                },
            ),
            Point(
                Coord {
                    lat: 53.0,
                    lng: 11.133333333333333,
                },
            ),
            Point(
                Coord {
                    lat: 53.005,
                    lng: 7.216666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 53.0425,
                    lng: 7.21,
                },
            ),
            Point(
                Coord {
                    lat: 53.08638888888889,
                    lng: 7.202222222222223,
                },
            ),
            Point(
                Coord {
                    lat: 53.105555555555554,
                    lng: 7.204444444444444,
                },
            ),
            Point(
                Coord {
                    lat: 53.113055555555555,
                    lng: 7.202777777777778,
                },
            ),
            Point(
                Coord {
                    lat: 53.13305555555556,
                    lng: 7.184722222222223,
                },
            ),
            Point(
                Coord {
                    lat: 53.13777777777778,
                    lng: 7.183888888888889,
                },
            ),
            Point(
                Coord {
                    lat: 53.157222222222224,
                    lng: 7.1925,
                },
            ),
            Point(
                Coord {
                    lat: 53.16361111111111,
                    lng: 7.196944444444445,
                },
            ),
            Point(
                Coord {
                    lat: 53.175555555555555,
                    lng: 7.2058333333333335,
                },
            ),
            Point(
                Coord {
                    lat: 53.17861111111111,
                    lng: 7.216388888888889,
                },
            ),
            Point(
                Coord {
                    lat: 53.18583333333333,
                    lng: 7.228055555555556,
                },
            ),
            Point(
                Coord {
                    lat: 53.18833333333333,
                    lng: 7.220833333333333,
                },
            ),
            Point(
                Coord {
                    lat: 53.19277777777778,
                    lng: 7.218888888888889,
                },
            ),
            Point(
                Coord {
                    lat: 53.20916666666667,
                    lng: 7.222777777777778,
                },
            ),
            Point(
                Coord {
                    lat: 53.22222222222222,
                    lng: 7.219444444444445,
                },
            ),
            Point(
                Coord {
                    lat: 53.225,
                    lng: 7.215277777777778,
                },
            ),
            Point(
                Coord {
                    lat: 53.23861111111111,
                    lng: 7.213333333333334,
                },
            ),
            Point(
                Coord {
                    lat: 53.238055555555555,
                    lng: 7.2155555555555555,
                },
            ),
            Point(
                Coord {
                    lat: 53.3,
                    lng: 7.191666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 53.31666666666667,
                    lng: 7.025,
                },
            ),
            Point(
                Coord {
                    lat: 53.36666666666667,
                    lng: 6.95,
                },
            ),
            Point(
                Coord {
                    lat: 53.46666666666667,
                    lng: 6.916666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 53.483333333333334,
                    lng: 6.891666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 53.50416666666667,
                    lng: 6.741666666666667,
                },
            ),
            Point(
                Coord {
                    lat: 53.516666666666666,
                    lng: 6.683333333333334,
                },
            ),
            Point(
                Coord {
                    lat: 53.63333333333333,
                    lng: 6.583333333333333,
                },
            ),
            Point(
                Coord {
                    lat: 53.666666666666664,
                    lng: 6.5,
                },
            ),
            Point(
                Coord {
                    lat: 55.0,
                    lng: 6.5,
                },
            ),
        ],
    },
    frequency: None,
    call_sign: None,
    transponder_code: None,
}
